src/main/java/de/tum/in/www1/artemis/service/compass/CompassService.java

```diff
@@ -221,68 +246,73 @@ public void addAssessment(long exerciseId, long submissionId, List<Feedback> mod
-                result.setResultString(points, modelingExercise.getMaxScore());
-
-                resultRepository.save(result);
-                engine.removeModelWaitingForAssessment(modelId, true);
-            }
-            else {
-                log.info("Model " + modelId + " got a confidence of " + grade.getConfidence() + " and a coverage of " + grade.getCoverage());
+        ModelingSubmission modelingSubmission = optionalModelingSubmission.get();
+
+        Result result = resultRepository.findDistinctWithFeedbackBySubmissionId(modelId)
+                .orElse(new Result().submission(modelingSubmission).participation(modelingSubmission.getParticipation()));
+        // only assess automatically when there is no manual assessment yet
+        if (result.getAssessmentType() != AssessmentType.MANUAL && result.getAssessor() == null) {
+            ModelingExercise modelingExercise = modelingExerciseRepository.findById(result.getParticipation().getExercise().getId())
+                    .orElseThrow(() -> new IllegalStateException("Exercise referenced in participation could not be found"));
+
+            // Workaround for ignoring automatic assessments of unsupported modeling exercise types TODO remove this after adapting compass
+            if (!isSupported(modelingExercise.getDiagramType())) {
+                return;
             }
+            // Round compass grades to avoid machine precision errors, make the grades more readable and give a slight advantage which makes 100% scores easier reachable
+            // see: https://confluencebruegge.in.tum.de/display/ArTEMiS/Feature+suggestions for more information
+            Grade grade = roundGrades(engine.getGradeForModel(modelId));
+
+            // Set feedback and assessment type of result
+            List<Feedback> automaticFeedbackAssessments = engine.convertToFeedback(grade, modelId, result);
+            result.getFeedbacks().clear();
+            result.getFeedbacks().addAll(automaticFeedbackAssessments);
+            result.setHasFeedback(false);
+            result.setAssessmentType(AssessmentType.AUTOMATIC);
+
+            saveResult(result, modelingSubmission);
         }
         else {
             // Make sure next optimal model is in a valid state
             engine.removeModelWaitingForAssessment(modelId, true);
         }
     }
 
+    /**
+     * Saves the given result to the database. If the result is new (i.e. no ID), the result is additionally assigned to the given submission which is then saved to the database as
+     * well.
+     *
+     * @param result             the result that should be saved
+     * @param modelingSubmission the corresponding modeling submission
+     */
+    private void saveResult(Result result, ModelingSubmission modelingSubmission) {
+        boolean isNewResult = result.getId() == null;
+
+        resultRepository.save(result);
+
+        if (isNewResult) {
+            modelingSubmission.setResult(result);
```

Comment:
If the result didn't have an id, it still doesn't have here.
I think there needs to be a new variable that stores the result of `resultRepository.save(result);`

Reply:
Good catch!
I am not sure if it makes any difference, but it's definitely cleaner. I changed it.